code/game/machinery/economy.dm

```diff
@@ -0,0 +1,406 @@
+/*
+Contains:
+
+	-ATMs
+	-Central banking server
+	-Bank accounts
+
+*/
+
+/*
+ * ATM's
+*/
+
+var/list/atms_in_world = list()
+/obj/machinery/atm
+	name = "automated transfer machine"
+	desc = "A Nanotrasen Bank of Credit ATM, used for transferring funds to and from ID cards."
+	icon = 'icons/obj/terminals.dmi'
+	icon_state = "dorm_available"
+	layer = 3
+	anchored = 1
+	var/obj/item/weapon/card/id/id = null //ID currently in the ATM
+	var/authenticated = FALSE // if the ID has been authenticated
+	var/obj/machinery/bankserver/linked_server = null //The server the ATM is controlled by
+	var/datum/html_interface/hi
+	var/datum/bankaccount/current_account
+	var/global/list/valid_screens = list("change_card_pin", "create_account", "switch_account", "change_account_pin", "withdraw", "deposit", "transfer")
+	var/current_screen = null
+
+/obj/machinery/atm/examine(mob/user)
+	..()
+	if(emagged)
+		user << "It appears to be in maintenance mode."
+	if(!linked_server)
+		user << "A small red LED is flashing."
+
+/obj/machinery/atm/New()
+	. = ..()
+
+	src.hi = new/datum/html_interface/nanotrasen(src, "Automated Teller Machine", 400, 600)
+
+	src.update()
+
+	atms_in_world.Add(src)
+	spawn(20)
+		linked_server = bankservers_in_world.len > 0 ? pick(bankservers_in_world) : null
+		if(!linked_server)
+			say("Notice: No banking servers detected to link to. Remote account access will be unavailable until this is resolved.")
+
+/obj/machinery/atm/Destroy()
+	qdel(src.hi)
+	src.hi = null
+
+	src.ejectID()
+	atms_in_world.Remove(src)
+	return ..()
+
+/obj/machinery/atm/proc/validatePIN(pin)
+	pin = text2num(pin)
+	return pin >= 1000 && pin <= 9999
+
+/obj/machinery/atm/Topic(href, href_list[], datum/html_interface_client/hclient)
+	if (istype(hclient))
```

Comment:
>unsanitized Topic()

you can literally steal the credits of anybody that puts their card into an ATM and route them to your account by exploiting the hrefs here.